### PR TITLE
`Quiz exercises`: Export images of quiz exercises

### PR DESCRIPTION
**Pull request by EneaGore**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/development-process.html#naming-conventions-for-github-pull-requests).
#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data.
- [x] I **strictly** followed the [client coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
~Following the [theming guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.~
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).
~I added `authorities` to all new routes and checked the course groups for displaying navigation elements (links, buttons).~
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
~I translated all newly inserted strings into English and German.~
### Motivation and Context

When a quiz is deleted, so are the assets. A previously imported json file of a quiz can no longer be used to import that quiz. This PR exports assets(images) in a zip file. 
Adress issue [7682](https://github.com/ls1intum/Artemis/issues/7682)
At the moment importing a quiz through a json file does not work if the images have been deleted. A follow up PR will allow to import the quiz with the exported zip folder and a json file.

### Description

For drag and drop questions the background and drag items are pulled and zipped. For images embedded in the text/MC Options a regex expression fetches the file name and path and zips the image.

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 quiz

1. Log in to Artemis
2. Create a Quiz with different type of questions
3. Add image in the text/ MC Options / Drag and Drop questions etc.
4. Export quiz
5. If there are no images a single json file of the quiz structure is downloaded
6. If there are images, a zip file with the images and a json file of the quiz structure is downloaded

[](https://artemis-test1.artemis.cit.tum.de)
[](https://artemis-test2.artemis.cit.tum.de)
[](https://artemis-test3.artemis.cit.tum.de)
[](https://artemis-test4.artemis.cit.tum.de)
[](https://artemis-test5.artemis.cit.tum.de)
[](https://artemis-test6.artemis.cit.tum.de)
[](https://artemis-test9.artemis.cit.tum.de)

#### Performance Review
- [x] I (as a reviewer) confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance 
- [x] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### Test Coverage

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|-----------------------------:|
| quiz-exercise.service.ts | 96.19% | ✅              |
### Screenshots

assetsInQuiz
jsonAndImage

## Summary by CodeRabbit

- **New Features**
  - Enhanced quiz export functionality to include images in markdown and assets from drag-and-drop exercises.
  - Added the ability to download a ZIP file containing exported quiz assets.

- **Tests**
  - Introduced new test cases to verify the export of images and assets from various quiz question types.

### CODE REVIEW

**Review by laadvo — APPROVED** [review]
Tested on ts4, the images (background and drag item) are correctly exported in a zip file

**Review by SimonEntholzer — APPROVED** [review]
Tested in testing session on ts2

**Review by DominikRemo — APPROVED** [review]
Tested in testing session on ts2. LGTM

**Review by kaancayli — DISMISSED** [review]
Tested on TS2 during testing session. Code looks good 👍 Works as expected.

File: src/main/webapp/app/exercises/quiz/manage/quiz-exercise.service.ts [hunk 1]
```diff
@@ -242,6 +246,62 @@ export class QuizExerciseService {
...
+    }
+
+    /**
+     * Exports assets (images) embedded in the markdown and from drag and drop exercises
+     * @param questions list of questions which will be exported
+     * @param fileName name of the output zip file
+     */
+    exportAssetsFromAllQuestions(questions: QuizQuestion[], fileName: string) {
+        const zip: JSZip = new JSZip();
+        const filePromises: any[] = [];
```
Comment (BBesrour) [inline src/main/webapp/app/exercises/quiz/manage/quiz-exercise.service.ts:259]:
```suggestion
        const filePromises: Promise<void>[] = [];
```

**Review by BBesrour — APPROVED** [review]
code lgtm!
small nitpick that could be ignored

**Review by coolchock — APPROVED** [review]
tested in testing session, LGTM

**Review by sarpsahinalp — APPROVED** [review]
Tested on ts2 works as expected.

**Review by FelixTJDietrich — APPROVED** [review]
Tested on TS6 also with edge cases and it worked in all cases. Code looks good overall.

**Review by coolchock — APPROVED** [review]
re-approve

**Review by kaancayli — APPROVED** [review]
Reapprove

**Review by FelixTJDietrich — APPROVED** [review]
Tested on TS6 and it is now one zip containing everything. Code changes also look good

image

**Review by marlon-luca-bu — APPROVED** [review]
Changes look good to me

**Pull request merged**